Prototype for m2e-pde source locator

laeubi opened the pull request.

laeubi commented:
Still blocked by
- https://github.com/eclipse-pde/eclipse.pde/pull/214

HannesWell commented on org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/MavenPluginSourcePathLocator.java:
```diff
@@ -0,0 +1,135 @@
+		ITargetPlatformService platformService = TargetPlatformService.getDefault();
+		try {
+			TargetBundle[] bundles = platformService.getWorkspaceTargetDefinition().getBundles();
+			for (TargetBundle targetBundle : bundles) {
```
Iterating over all bundles in the TP could be a relatively slow operation if there are many.
It would be ideal if there would be some mean to obtain the TargetBundle of a external Plugin model, but I didn't find any. So maybe it would be reasonable to maintain a corresponding Mapping within M2E that is updated when the TP changes.

HannesWell reviewed (changes requested):
In general it works but needs some adjustments to work with artifacts with inherited groupId+ version.
And we should have a look at some potential/assumed performance issues.

HannesWell commented:
This change should use the `MavenArtifactIdentifier.identify()` and `MavenArtifactIdentifier.resolveSourceLocation()` methods.

HannesWell commented:
Eclipse 4.26 is now available in the TP.

laeubi commented on org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/MavenPluginSourcePathLocator.java:
```diff
@@ -0,0 +1,135 @@
+				String name = jarEntry.getName();
+				if (name.startsWith("META-INF/maven/")) {
+					if (name.endsWith("/pom.properties")) {
+						try (InputStream stream = jarFile.getInputStream(jarEntry)) {
```
It all depends how you define "faster"... reading the xml might take a few milliseconds more but this actually don't matter much here...

laeubi commented on org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/MavenPluginSourcePathLocator.java:
```diff
@@ -0,0 +1,135 @@
+	private Artifact recoverFromXML(InputStream stream) {
+		try {
+			Model model = MavenPlugin.getMavenModelManager().readMavenModel(stream);
+			return new DefaultArtifact(model.getGroupId(), model.getArtifactId(), null, model.getVersion());
```
There is no way to read the parent from an embedded pom, such poms should be "flatten" already.

laeubi commented:
> Eclipse 4.26 is now available in the TP.

I updated the PR and added a reverse file lookup, that should be "fast enough"... especially as in my testing in almost all cases the sources are already found beforehands.

laeubi commented:
> This change should use the MavenArtifactIdentifier.identify() and MavenArtifactIdentifier.resolveSourceLocation() methods.

Done.

HannesWell commented:
Looks good.

> > Eclipse 4.26 is now available in the TP.
> 
> I updated the PR and added a reverse file lookup, that should be "fast enough"... especially as in my testing in almost all cases the sources are already found beforehands.

Are the sources also resolved during the target-resolution? That would maybe explain it.
However in cases where a artifact from Maven-Central is re-distributed via a P2-repo that does not contain the sources, this is still useful.

laeubi commented:
Depends on

1. include source enabled in the target
2. one has already run a maven build that uses the artifact and sources are downloaded
